### PR TITLE
BUG: Series.groupby fails with InvalidIndexError on time series with …

### LINKED ISSUE #42731: BUG: Series.groupby fails with InvalidIndexError on time series with a tuple-named grouper.
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

s = pd.Series(index=[pd.Timestamp(2021,7,26)], name=('A', 1))
s.groupby(s==s)  # raises InvalidIndexError
```

#### Problem description

There are several things necessary to trigger this:
- The index has to consist of `pd.Timestamp`s. If you change the example to `s = pd.Series(index=[0,1], name=('A', 1))` everything is fine.
- The grouper, i.e. `groupby`'s `by` argument , has to be a `Series` named with a tuple (like you get if you select a `MultiIndex`-column from a `DataFrame`). If you change the example to `s.groupby((s==s).rename('foo'))` everything is fine. Same goes if you use a pure list or a function.
- The object to group must be a `Series`. If you change the example to `pd.DataFrame(s).groupby(s==s)` everything is fine.
- You need `pandas>=1.1.3`. If you downgrade to an older version - guess what - everything is fine.

Why this is a problem: The name of a grouper should not matter and it is just not to understand why a time series behaves differently than any other kind of `Series`.

#### Expected Output

Behavior should be as for non-time-series.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.8.3.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : de_DE.cp1252
pandas           : 1.1.3
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.1.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.2
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by ghost**
…a tuple-named grouper. (#42731)

- [x] closes #42731
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Skipped any "What's New" for now, since it's not decided which version will have it. Will update as soon as this is decided.

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -845,7 +845,7 @@ def is_in_obj(gpr) -> bool:
             return False
         try:
             return gpr is obj[gpr.name]
-        except (KeyError, IndexError):
+        except (KeyError, IndexError, InvalidIndexError):
             # IndexError reached in e.g. test_skip_group_keys when we pass
             #  lambda here
```
Comment (jbrockmendel) [inline pandas/core/groupby/grouper.py:850]: can you extend this comment to explain InvalidIndexError

Comment (suoniq) [reply pandas/core/groupby/grouper.py:850]: Done.

**Comment by suoniq** [comment]
Please note that I'll be on vacation until mid-August and plan to not come close to any computers. So I won't respond Do any necessary changes to the MR if you don't want to wait. See you then.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note for 1.4, grouping section in bug fixes

**Comment by jreback** [comment]
cc @rhshadrach if you can look

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -2311,6 +2311,15 @@ def test_groupby_series_with_tuple_name():
     tm.assert_series_equal(result, expected)
 
 
+def test_time_series_groupby_series_named_with_tuple():
+    # GH 42731
+    ser = Series([1, 2, 3, 4], index=[Timestamp(2021, 7, 28 + i) for i in range(4)])
+    grp = Series([1, 0, 1, 0], name=('a', 'a'))
+    result = ser.groupby(grp).last()
+    expected = ser.groupby(grp.rename(None)).last()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:2319]: Passing a Series into the groupby will try to align, e.g.

```
ser = Series([1, 2, 3, 4], index=[3, 1, 2, 0])
grp = Series([1, 0, 1, 0])
print(ser.groupby(grp).sum())
```

gives the result

```
0    3
1    7
dtype: int64
```

In this case - the index won't align and you end up with an empty result, is that right?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Fix looks good - but I don't think the test is as intended.

**Comment by rhshadrach** [comment]
@suoniq - friendly ping. It'd be great to get this in for 1.3.2 and I think it's almost there. If you don't have time to finish it up, would you mind if I pushed it across the finish line?

**Comment by rhshadrach** [comment]
@simonjayhawkins - I've tagged this as 1.3.2 but certainly don't hold up the release if it doesn't get in by tomorrow.

@suoniq - Thanks for the PR and apologies for the ping above, I missed your note about being out.

cc @jreback @jbrockmendel

**Comment by simonjayhawkins** [comment]
moving to 1.3.3

**Comment by suoniq** [comment]
Hi, back from vacation.  
@rhshadrach Of course I am fine with you taking over. Still consider the original test as good b/c it pin-pointed exactly the problem: time series behave differently than series indexed differently. Was red before the fix as well. `(-;`  
Anyway, please let me know, if there is anything left I can do.

**Comment by rhshadrach** [comment]
@suoniq I didn't think the intention was to test on empty objects, but let me know if this is incorrect. I think we're in good shape here, just waiting on review. More than happy if you want to take the reins back again!

**Comment by rhshadrach** [comment]
@jreback and @jbrockmendel for review

File: pandas/tests/groupby/test_grouping.py [hunk 3]
```diff
@@ -396,6 +396,24 @@ def test_groupby_dict_mapping(self):
...
+    @pytest.mark.parametrize(
+        "index",
+        [
+            [0, 1, 2, 3],
+            ["a", "b", "c", "d"],
+            [Timestamp(2021, 7, 28 + i) for i in range(4)],
+        ],
+    )
+    @pytest.mark.parametrize("box", [Series, DataFrame])
+    def test_groupby_series_named_with_tuple(self, box, index):
```
Comment (jreback) [inline pandas/tests/groupby/test_grouping.py:408]: use `frame_or_series` fixture rather than box

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment, pls merge master and ping on green

**Comment by rhshadrach** [comment]
Thanks @jreback - comment addressed and green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @suoniq and @rhshadrach

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**